File: catkin_ws/src/abb_robot_sandbox/exec/egm_controller_node.py
#!/usr/bin/env python

import rospy
import sys
from sensor_msgs.msg import JointState
import abb_robot_sandbox
from abb_robot_sandbox.msg import ControlGoal, ControlFeedback, ControlResult
import actionlib
from abb_robot_sandbox.util import RobotKinematics
from visualization_msgs.msg import Marker
from abb_robot_msgs.srv import TriggerWithResultCode
from std_msgs.msg import Float64MultiArray
from controller_manager_msgs.srv import SwitchController
from abb_rapid_sm_addin_msgs.srv import GetEGMSettings, SetEGMSettings
from abb_egm_msgs.msg import EGMState
import math


class EGMController:
    def __init__(self, name):
        self._action_server = actionlib.SimpleActionServer(
            name,
            abb_robot_sandbox.msg.ControlAction,
            execute_cb=self.callback,
            auto_start=False)
        self._action_server.start()

    def init(self):
        if not rospy.has_param('~freq'):
            return False
        self._freq = rospy.get_param('~freq')

        if not rospy.has_param('~topics/joint_states_in'):
            return False
        self._joint_states_pub = rospy.Publisher(
            rospy.get_param('~topics/joint_states_in'), JointState, queue_size=15)

        # RWS services.
        self._stop_rapid_srv = rospy.ServiceProxy(
            '/rws/stop_rapid', TriggerWithResultCode)
        self._pp_to_main_srv = rospy.ServiceProxy(
            '/rws/pp_to_main', TriggerWithResultCode)
        self._start_rapid_srv = rospy.ServiceProxy(
            '/rws/start_rapid', TriggerWithResultCode)
        self._start_egm_joint_srv = rospy.ServiceProxy(
            '/rws/sm_addin/start_egm_joint', TriggerWithResultCode)
        self._stop_egm_srv = rospy.ServiceProxy(
            '/rws/sm_addin/stop_egm', TriggerWithResultCode)

        self._switch_controller_srv = rospy.ServiceProxy(
            '/egm/controller_manager/switch_controller', SwitchController)

        self._vel_cmd_pub = rospy.Publisher(
            '/egm/joint_group_velocity_controller/command',
            Float64MultiArray, queue_size=10)
        self._pos_cmd_pub = rospy.Publisher(
            '/egm/joint_group_position_controller/command',
            Float64MultiArray, queue_size=10)

        self._get_egm_settings_srv = rospy.ServiceProxy(
            "/rws/sm_addin/get_egm_settings", GetEGMSettings)
        self._set_egm_settings_srv = rospy.ServiceProxy(
            "/rws/sm_addin/set_egm_settings", SetEGMSettings)

        self._egm_state_sub = rospy.Subscriber(
            '/egm/egm_states', EGMState, self._egm_state_callback)
        self._egm_joint_state_sub = rospy.Subscriber(
            '/egm/joint_states', JointState, self._egm_joint_state_callback)
        self._egm_state = None
        self._egm_joint_state = None

        # For visualizing the desired EE position.
        if not rospy.has_param('~robot_base_frame_id'):
            return False
        self._robot_base_frame_id = rospy.get_param(
            '~robot_base_frame_id')

        if not rospy.has_param('~robot_tip_frame_id'):
            return False
        self._robot_tip_frame_id = rospy.get_param(
            '~robot_tip_frame_id')

        # Initialize the robot kinematics (FK/IK).
        self._kinematics = RobotKinematics(
            self._robot_base_frame_id, self._robot_tip_frame_id)

        if not rospy.has_param('~topics/vis'):
            return False
        vis_topic = rospy.get_param('~topics/vis')
        self._vis_pub = rospy.Publisher(vis_topic, Marker, queue_size=25)

        if not rospy.has_param('~joint_pos_tol'):
            return False
        self._joint_pos_tol = rospy.get_param('~joint_pos_tol')

        return True

    def _egm_state_callback(self, msg):
        self._egm_state = msg

    def _egm_joint_state_callback(self, msg):
        self._egm_joint_state = msg

    def start_egm_session(self):
        current_settings = self._get_egm_settings_srv(task='T_ROB1')
        settings = current_settings.settings
        print(settings)

        settings.activate.max_speed_deviation = math.degrees(3.0)
        print(settings)
        self._set_egm_settings_srv(task='T_ROB1', settings=settings)

        rospy.loginfo("[Controller] Stopping RAPID...")
        res = self._stop_rapid_srv()
        if res.result_code != 1:
            rospy.logerr("[Controller] {}".format(res.message))
            # return False

        rospy.sleep(2)

        rospy.loginfo("[Controller] Setting program pointer to main...")
        res = self._pp_to_main_srv()
        if res.result_code != 1:
            rospy.logerr("[Controller] {}".format(res.message))
            # return False

        rospy.sleep(2)

        rospy.loginfo("[Controller] Starting RAPID...")
        res = self._start_rapid_srv()
        if res.result_code != 1:
            rospy.logerr("[Controller] {}".format(res.message))
            # return False

        rospy.sleep(2)

        rospy.loginfo("[Controller] Starting EGM session...")
        res = self._start_egm_joint_srv()
        if res.result_code != 1:
            rospy.logerr("[Controller] {}".format(res.message))
            return False

        rospy.sleep(0.5)

        rospy.loginfo("[Controller] ...done!")
        return True

    def stop_egm_session(self):
        rospy.loginfo("[Controller] Stopping EGM...")
        res = self._stop_egm_srv()
        if res.result_code != 1:
            rospy.logerr("[Controller] {}".format(res.message))
            return False

        rospy.loginfo("[Controller] ...done")
        return True

    def start_velocity_controller(self):
        rospy.loginfo(
            "[Controller] Switching to joint group velocity controller...")
        res = self._switch_controller_srv(
            ['joint_group_velocity_controller'], [], 1, False, 0.0)
        if not res.ok:
            return False

        rospy.loginfo("[Controller] ...done")
        return True

    def start_position_controller(self):
        rospy.loginfo(
            "[Controller] Switching to joint group position controller...")
        res = self._switch_controller_srv(
            ['joint_group_position_controller'], [], 1, False, 0.0)
        if not res.ok:
            return False

        rospy.loginfo("[Controller] ...done")
        return True

    def callback(self, goal):
        rospy.loginfo("[Controller] Received trajectory with {} waypoints".
                      format(len(goal.trajectory.points)))

        rate = rospy.Rate(self._freq)

        feedback = ControlFeedback()
        time = rospy.Time.now()
        success = True

        self._egm_state = None
        self._egm_joint_state = None

        index = 0

        if not self.start_egm_session():
            result = ControlResult()
            result.result = ControlResult.FAILURE
            self._action_server.set_succeeded(result)
            return

        # if not self.start_velocity_controller():
        if not self.start_position_controller():
            result = ControlResult()
            result.result = ControlResult.FAILURE
            self._action_server.set_succeeded(result)
            return

        while not rospy.is_shutdown():
            if self._action_server.is_preempt_requested():
                rospy.logwarn("[Controller] Preempt requested!")
                self._action_server_.set_preempted()
                success = False
                break

            if self._egm_state is None:
                rospy.logwarn("[Controller] Waiting for EGM state message...")
                rate.sleep()
                continue
            else:
                if len(self._egm_state.egm_channels) == 0 or \
                   not self._egm_state.egm_channels[0].active:
                    rospy.logwarn("[Controller] EGM is not active!")
                    success = False
                    break

            if self._egm_joint_state is None:
                rospy.logwarn(
                    "[Controller] Waiting for EGM joint state message...")
                rate.sleep()
                continue

            advance_index = True
            for i in range(6):
                if abs(goal.trajectory.points[index].positions[i] -
                       self._egm_joint_state.position[i]) > self._joint_pos_tol:
                    advance_index = False
                    break

            if advance_index:
                index += 1

            if index >= len(goal.trajectory.points):
                rospy.loginfo("[Controller] Reached end of trajectory!")
                break

            # time_from_start = rospy.Time.now() - time

            # if goal.trajectory.points[-1].time_from_start <= time_from_start:
            #     rospy.loginfo("[Controller] Reached end of trajectory!")
            #     break

            # index = -1
            # for i in range(len(goal.trajectory.points) - 1):
            #     point = goal.trajectory.points[i]
            #     next_point = goal.trajectory.points[i + 1]
            #     if point.time_from_start < time_from_start and \
            #        time_from_start <= next_point.time_from_start:
            #         index = i
            #         break

            rospy.loginfo("[Controller] At index {}".format(index))
            if index < 0:
                success = False
                break

            # Feedback gives the current desired joint config.
            joint_states_msg = JointState()
            joint_states_msg.position = goal.trajectory.points[index].positions
            # self._joint_states_pub.publish(joint_states_msg)

            # Publish the joint position commands to EGM.
            pos_cmd_msg = Float64MultiArray()
            pos_cmd_msg.data = goal.trajectory.points[index].positions
            self._pos_cmd_pub.publish(pos_cmd_msg)

            self.pub_ee_marker(pos_cmd_msg.data)

            feedback.config = joint_states_msg
            self._action_server.publish_feedback(feedback)

            rate.sleep()

        result = ControlResult()
        if success:
            result.result = ControlResult.SUCCESS
        else:
            result.result = ControlResult.FAILURE

        self.stop_egm_session()

        self._action_server.set_succeeded(result)

    def pub_ee_marker(self, config):
        msg = Marker()
        msg.header.frame_id = self._robot_base_frame_id
        msg.id = 0
        msg.type = Marker.SPHERE
        msg.ns = 'controller/ee'
        msg.scale.x = 0.1
        msg.scale.y = 0.1
        msg.scale.z = 0.1
        msg.color.r = 0.5
        msg.color.g = 0.5
        msg.color.b = 0
        msg.color.a = 0.75

        pos, _ = self._kinematics.fk(config)
        msg.pose.position.x = pos[0]
        msg.pose.position.y = pos[1]
        msg.pose.position.z = pos[2]
        msg.pose.orientation.x = 0
        msg.pose.orientation.y = 0
        msg.pose.orientation.z = 0
        msg.pose.orientation.w = 1

        self._vis_pub.publish(msg)

    def run(self):
        rospy.spin()


def main():
    rospy.init_node('egm_controller')

    controller = EGMController(rospy.get_name())
    if not controller.init():
        rospy.logerr("[EGMController] Error: failed to initialize!")
        sys.exit(-1)

    controller.run()


if __name__ == '__main__':
    main()

File: catkin_ws/src/abb_robot_sandbox/exec/egm_controller_node.cpp
#include <abb_libegm/egm_controller_interface.h>
#include <abb_robot_sandbox/ControlAction.h>
#include <actionlib/server/simple_action_server.h>
#include <angles/angles.h>
#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <trajectory_msgs/JointTrajectory.h>

class EGMController {
public:
  EGMController()
      : p_nh_("~"), freq_(30.0), egm_port_(6511), egm_rate_(250.0),
        action_server_(nh_, "controller",
                       boost::bind(&EGMController::ExecuteCallback, this, _1),
                       false) {
    action_server_.start();
  }

  ~EGMController() {
    if (egm_interface_) {
      delete egm_interface_;
      // egm_interface_ = nullptr;
      egm_interface_ = 0;
    }

    io_service_.stop();
    thread_group_.join_all();
  }

  bool Init() {
    if (!p_nh_.getParam("freq", freq_))
      return false;

    if (!p_nh_.getParam("egm/port", egm_port_))
      return false;

    if (!p_nh_.getParam("egm/rate", egm_rate_))
      return false;

    ROS_INFO_STREAM("[EGMController] Initializing EGM interface on port "
                    << egm_port_ << "...");
    egm_interface_ =
        new abb::egm::EGMControllerInterface(io_service_, egm_port_);

    if (!egm_interface_->isInitialized()) {
      ROS_ERROR_STREAM("[EGMController] EGM interface initializion failed "
                       "(port already bound?)");
      return false;
    }

    thread_group_.create_thread(
        boost::bind(&boost::asio::io_service::run, &io_service_));
    ROS_INFO_STREAM("...done!");

    std::string joint_states_topic;
    if (!p_nh_.getParam("topics/joint_states", joint_states_topic))
      return false;
    joints_pub_ =
        nh_.advertise<sensor_msgs::JointState>(joint_states_topic, 1000);

    return true;
  }

  void ExecuteCallback(const actionlib::SimpleActionServer<
                       abb_robot_sandbox::ControlAction>::GoalConstPtr &goal) {
    // Wait for EGM communcations to start.
    ROS_INFO("[EGMController] Waiting for EGM communcations...");
    bool wait = true;
    while (ros::ok() && wait) {
      if (action_server_.isPreemptRequested()) {
        ROS_WARN("[EGMController] Preempted!");
        action_server_.setPreempted();
        return;
      }

      if (egm_interface_->isConnected()) {
        if (egm_interface_->getStatus().rapid_execution_state() ==
            abb::egm::wrapper::Status_RAPIDExecutionState_RAPID_UNDEFINED) {
          ROS_WARN("[EGMController] RAPID execution state is UNDEFINED!");
        } else {
          wait = egm_interface_->getStatus().rapid_execution_state() !=
                 abb::egm::wrapper::Status_RAPIDExecutionState_RAPID_RUNNING;
        }
      }

      ros::Duration(0.5).sleep();
    }

    // Execute the trajectory.
    abb::egm::wrapper::Input input;
    abb::egm::wrapper::Output output;
    int sequence_number = 0;
    int first_sequence_number = -1;
    double time = 0.0;
    bool success = true;
    while (ros::ok()) {
      if (action_server_.isPreemptRequested()) {
        ROS_WARN("[EGMController] Preempted!");
        action_server_.setPreempted();
        success = false;
        break;
      }

      if (egm_interface_->waitForMessage(500)) {
        egm_interface_->read(&input);
        sequence_number = input.header().sequence_number();

        if (first_sequence_number < 0) {
          first_sequence_number = sequence_number;
          ROS_INFO_STREAM("[EGMController] First sequence number is "
                          << first_sequence_number);
        }

        time = static_cast<double>(sequence_number - first_sequence_number) /
               egm_rate_;

        if (goal->trajectory.points.back().time_from_start.toSec() <= time) {
          ROS_INFO("[EGMController] Reached end of trajectory!");
          break;
        }

        // Get the waypoint to send to robot.
        int index = -1;
        for (int i = 0; i < goal->trajectory.points.size() - 1; ++i) {
          const trajectory_msgs::JointTrajectoryPoint &point =
              goal->trajectory.points[i];
          const trajectory_msgs::JointTrajectoryPoint &next_point =
              goal->trajectory.points[i + 1];
          if (point.time_from_start.toSec() <= time &&
              time < next_point.time_from_start.toSec()) {
            index = i;
            break;
          }
        }

        if (index < 0) {
          ROS_WARN("[EGMController] Failed to get waypoint!");
          success = false;
          break;
        }

        const trajectory_msgs::JointTrajectoryPoint &point =
            goal->trajectory.points[index];

        output.mutable_robot()->mutable_joints()->mutable_position()->CopyFrom(
            input.feedback().robot().joints().position());
        for (int i = 0; i < 6; ++i) {
          output.mutable_robot()
              ->mutable_joints()
              ->mutable_position()
              ->set_values(i, angles::to_degrees(point.positions[i]));
        }

        egm_interface_->write(output);
      }
    }

    ROS_INFO("[EGMController] Trajectory execution complete!");
    abb_robot_sandbox::ControlResult result;
    if (success)
      result.result = abb_robot_sandbox::ControlResult::SUCCESS;
    else
      result.result = abb_robot_sandbox::ControlResult::FAILURE;

    action_server_.setSucceeded(result);
  }

  void Run() {
    sensor_msgs::JointState joints_msg;
    joints_msg.header.seq = 0;
    for (int i = 0; i < 6; ++i) {
      joints_msg.name.push_back("joint_" + boost::to_string(i + 1));
      joints_msg.position.push_back(0);
      joints_msg.velocity.push_back(0);
      joints_msg.effort.push_back(0);
    }

    ros::Rate rate(freq_);
    while (ros::ok()) {
      // TODO Update joints?

      joints_msg.header.stamp = ros::Time::now();
      joints_msg.header.seq += 1;
      joints_pub_.publish(joints_msg);
      ros::spinOnce();
      rate.sleep();
    }
  }

private:
  ros::NodeHandle nh_;
  ros::NodeHandle p_nh_;
  double freq_;
  int egm_port_;
  double egm_rate_;
  ros::Publisher joints_pub_;

  actionlib::SimpleActionServer<abb_robot_sandbox::ControlAction>
      action_server_;

  boost::asio::io_service io_service_;
  boost::thread_group thread_group_;
  abb::egm::EGMControllerInterface *egm_interface_;
  int seq_num_;
};

int main(int argc, char *argv[]) {
  ros::init(argc, argv, "egm_controller");

  EGMController controller;
  if (!controller.Init()) {
    ROS_ERROR("[EGMController] Failed to initialize node!");
    return -1;
  }

  controller.Run();

  return 0;
}
